src/htm/engine/RESTapi.cpp

```diff
@@ -36,44 +37,45 @@ RESTapi::~RESTapi() { }
 
 RESTapi* RESTapi::getInstance() { return &rest; }
 
-std::string RESTapi::get_new_id_(const std::string &old_id) {
-  std::string id = old_id;
-
-  // re-use the same id if we can or use a provide one.
+  std::string RESTapi::get_new_id_() {
+  // No id was provided so find the next available number.
+  // Note: This will return a 4 digit number as a string, 
+  //       starting with "0001" and incrementing on each use. 
+  //       This will never return "0000" 
+  
   std::map<std::string, ResourceContext>::iterator itr;
-  if (id.empty()) {
-    // id is empty, create a new context id
-    while (rest.resource_.size() < UINT16_MAX - 1) {
-      unsigned int id_nbr = next_id++;
-      if (id_nbr == 0)
-        id_nbr = next_id++; // allow integer wrap of the id without a 0 value.
-      char buf[10];
-      std::snprintf(buf, sizeof(buf), "%4.04x", id_nbr);
-      id = buf;
-
-      // Make sure this new session id is not in use.
-      itr = resource_.find(id);
-      if (itr == resource_.end() || (itr->second.t < time(0) - RESOURCE_TIMEOUT)) {
-        // This is one we can use
-        break;
-      }
-    };
+  std::string id;
+  while (rest.resource_.size() < ID_MAX) {  // limit the total number of generated resources
+    unsigned int id_nbr = next_id++;
+    if (id_nbr > ID_MAX)
+      id_nbr = 1; // allow integer wrap of the id without a "0000" value.
+    char buf[10];
+    std::snprintf(buf, sizeof(buf), "%4.04x", id_nbr);
+    id = buf;
+
+    // Make sure this new session id is not in use.
+    itr = resource_.find(id);
+    if (itr == resource_.end() || (itr->second.t < time(0) - RESOURCE_TIMEOUT)) {
```

Comment:
there's a sign-compare issue, but looking at this, is the "resource timeout" a good thing? I am worried it can bring more (hidden) problems in production. 

Say I setup a long running (multiple days) task. The `RESOURCE_TIMEOUT` is 1 day. 
If I request a new ID after a day of training, it'll give me an "reused" id for some part of the network, and I'll break the Network. 
- do we need ID reuse, resource-timeout at all? 
- what if we just returned a new (random?) ID from the pool ([1, 9999])?

Reply:
That timer gets reset on each access of the Network object.   So to be a problem a user would not have accessed the Network object for an entire day.

The objective was to clean up abandoned Network objects.  
Lets say we have a  public REST server running all of the time and various people access it  to do their experiments.  Not all of them (if any) would think to delete their Network resource when they are done.  So this is sort of a background garbage collection.

If we don't collect the garbage the server could eventually run out of RAM.  

Perhaps we could require some sort of user login and tag all Network resources with the userid.  Then we could time-out the user, freeing up all of the resources.

Or maybe this is overkill unless we are really going to setup a public REST server.

Reply:
> So to be a problem a user would not have accessed the Network object for an entire day. 

I believe that is possible. Setup a long training dataset, create the network over REST, start training, wait a few days to get "learned state". So overall, I'm :+1: for the removal of the timeout that would kill the task. 

> ets say we have a public REST server running all of the time and various people access it to do their experiments. Not all of them (if any) would think to delete their Network resource when they are done. 

Ok, I misunderstood the intended usage. A public, multi-user REST HTM server... I think given the popularity of Docker and ondemand cloud resources, more likely use would be that each user spins their own, private REST server (easier resource & load management) which they control over the network.